.gitlab-ci.yml

```diff
@@ -31,57 +30,23 @@ default:
       - unknown_failure
       - api_failure
 
-.common-refs:                      &common-refs
-  rules:
-    - if: $BUILD
-      when: never
-    - if: $CI_COMMIT_REF_NAME == "master"
-    - if: $CI_COMMIT_REF_NAME =~ /^[0-9]+$/                         # PRs
-
 .test-refs:                        &test-refs
   rules:
-    - if: $CI_COMMIT_REF_NAME =~ /^[0-9]+$/                         # PRs
+    - if: $CI_COMMIT_REF_NAME =~ /^[0-9]+$/                 # PRs
 
-# to create, publish and deploy image
-.staging-refs:                     &staging-refs
+.publish-refs:                     &publish-refs
   rules:
     - if: $CI_COMMIT_REF_NAME == "master"
-    - if: $CI_COMMIT_REF_NAME =~ /^v[0-9]+\.[0-9]+.*$/             # i.e. v1.0, v2.1rc1
-    - if: $CI_COMMIT_REF_NAME =~ /^stg-v[0-9]+\.[0-9]+.*$/         # i.e. stg-v1.0, stg-v2.1rc1
 
-.production-refs:                  &production-refs
+.common-refs:                      &common-refs
   rules:
-    - if: $CI_COMMIT_REF_NAME =~ /^v[0-9]+\.[0-9]+.*$/             # i.e. v1.0, v2.1rc1
+    - !reference [.test-refs, rules]
+    - !reference [.publish-refs, rules]
 
-.master-refs:                      &master-refs
+.production-refs:                  &production-refs
   rules:
-    - if: $BUILD
-      when: never
     - if: $CI_COMMIT_REF_NAME == "master"
-
-# test that docker image can be built
-.build-only-docker-image:          &build-only-docker-image
-  image:                           $BUILDAH_IMAGE
-  script:
-    - buildah bud
-      --format=docker
-      --target="$TARGET"
-      --tag "$IMAGE_NAME:$DOCKER_TAG"
-      --file "$DOCKERFILE" .
-
-.build-push-docker-image:          &build-push-docker-image
-  image:                           $BUILDAH_IMAGE
-  script:
-    - echo "$Docker_Hub_Pass_Parity" |
-      buildah login --username "$Docker_Hub_User_Parity" --password-stdin docker.io
-    - buildah bud
-      --format=docker
-      --target=$TARGET
-      --tag "$IMAGE_NAME:$DOCKER_TAG"
-      --file $DOCKERFILE .
-    - buildah push --format=v2s2 "$IMAGE_NAME:$DOCKER_TAG"
-  after_script:
-    - buildah logout --all
+      when: manual
```

Comment:
i don't fully get why the prod refs are `master` + manual? could you please tech me how to read it ?

Reply:
This means that you can deploy to prod on master branch but you have to press the "deploy" button

Reply:
but that's not what we agreed to :) 

prod should be deployable by tagging new version `/^v[0-9]+\.[0-9]+.*$/             # i.e. v1.0, v2.1rc1` 
staging  either `/^stg-v[0-9]+\.[0-9]+.*$/` or master

now it looks like it's prod manually through gitlab and staging automatically from master, but sometimes I need to check some dev branch on stage, so only master doens't work

Reply:
yes, i think what i just wrote above corresponds to your suggestion

- ✅  automatic deployments from master to staging on merge (+ optionally manual ones, so we can test dev-branches for some apps)
- ✅ production deployment requiring a manual trigger (tagging or using github releases is a manual trigger. it also generates automatically the changelog)

Reply:
i thought when you mention "creating special tags" you mean how it was in the past, that staging was allowed only when you tag, but not from master. So many projects recently had a "master" ref for staging as well, which were auto-deploying the **staging**

for **prod** - we could also deploy it from button, but that wouldn't give a sense of what was releasing and when, so I think using GH releases with history of changelog is better option. At least, if not GH releases, then tagging would give a minimum sense of "when" and "what" you can infer from comparing diffs from tags.
So when you were saying "requiring a manual trigger" I didn't think about a button necessary, but GH releases, which is few clicks, pretty convenient and controlled